### ISSUE TITLE
[BUG] goodhosts cannot modify /etc/hosts despite being a setuid program

### ISSUE THREAD

**Issue by Christopher-Chianelli** [code-ready/crc#1425]
### General information

  * OS: Fedora 32
  * Hypervisor: KVM 
  * Did you run `crc setup` before starting it (Yes)?
  * Running CRC on:  Personal Computer

## CRC version
```bash
CodeReady Containers version: 1.13.0+8070bae
OpenShift version: 4.5.1 (embedded in binary)
```
  
## CRC status
```bash
# Put `crc status --log-level debug` output here
```

## CRC config
```bash
```
(No output)

## Host Operating System
```bash
NAME=Fedora
VERSION="32 (Workstation Edition)"
ID=fedora
VERSION_ID=32
VERSION_CODENAME=""
PLATFORM_ID="platform:f32"
PRETTY_NAME="Fedora 32 (Workstation Edition)"
ANSI_COLOR="0;34"
LOGO=fedora-logo-icon
CPE_NAME="cpe:/o:fedoraproject:fedora:32"
HOME_URL="https://fedoraproject.org/"
DOCUMENTATION_URL="https://docs.fedoraproject.org/en-US/fedora/f32/system-administrators-guide/"
SUPPORT_URL="https://fedoraproject.org/wiki/Communicating_and_getting_help"
BUG_REPORT_URL="https://bugzilla.redhat.com/"
REDHAT_BUGZILLA_PRODUCT="Fedora"
REDHAT_BUGZILLA_PRODUCT_VERSION=32
REDHAT_SUPPORT_PRODUCT="Fedora"
REDHAT_SUPPORT_PRODUCT_VERSION=32
PRIVACY_POLICY_URL="https://fedoraproject.org/wiki/Legal:PrivacyPolicy"
VARIANT="Workstation Edition"
VARIANT_ID=workstation

```

### Steps to reproduce

  1. crc setup
  2. crc start

### Expected

CRC host entry to be added to `/etc/hosts` via goodhosts

### Actual

goodhosts fails with "Host file not writable. Try running with elevated privileges."

### Logs

Before gather the logs try following if that fix your issue
```bash
Using SSH client type: external
Using SSH private key: /home/cchianel/.crc/cache/crc_libvirt_4.5.1/id_rsa_crc (-r--------)
&{[-F /dev/null -o ConnectionAttempts=3 -o ConnectTimeout=10 -o ControlMaster=no -o ControlPath=none -o LogLevel=quiet -o PasswordAuthentication=no -o ServerAliveInterval=60 -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null core@192.168.130.11 -o IdentitiesOnly=yes -i /home/cchianel/.crc/cache/crc_libvirt_4.5.1/id_rsa_crc -p 22] /usr/bin/ssh <nil>}
DEBU About to run SSH command:
sudo podman run  --ip 10.88.0.8 --name dnsmasq -v /var/srv/dnsmasq.conf:/etc/dnsmasq.conf -p 53:53/udp --privileged -d quay.io/crcont/dnsmasq:latest 
DEBU SSH command results: err: <nil>, output: e7d8b1f1b4922a8765a9be761c4aa3034786e8c41da92d5fc5ab55ec709b4e1b 
DEBU Running '/home/cchianel/.crc/bin/goodhosts rm api.crc.testing' 
DEBU Command failed: exit status 1                
DEBU stdout:                                      
DEBU stderr: Host file not writable. Try running with elevated privileges. 
ERRO Error running post start: exit status 1   
```

### Workaround

Before `crc start`, run
`sudo chmod 666 /etc/hosts` (read/write to all)

After `crc start`, run
`sudo chmod 644 /etc/hosts` (read all, write only to root).

Note:
permissions on goodhosts:
```bash
ls -l /home/cchianel/.crc/bin/goodhosts
-r-s--x---. 1 root cchianel 5551059 Jul 28 11:12 /home/cchianel/.crc/bin/goodhosts
```
Note it a setuid program, so it should be able to modify `/etc/hosts` with `/etc/hosts` permissions being `rw-r--r-- root root`

**Comment by praveenkumar** [comment]
@Christopher-Chianelli by default in the linux `/etc/hosts` file have `0644` permission so it might be you or some other application change it's permission and make is non writeable even for root users, did you check it again when set the file permission back to `0644` ?

**Comment by cfergeau** [comment]
goodhosts is supposed to be a fallback in case name resolution is not working as expected, maybe its failures should not be fatal?

**Comment by Christopher-Chianelli** [comment]
When I checked (before chmoding anything), /etc/hosts is readable/writeable for root, read-only for non-root (0644 root root). `goodhosts` fails with these permissions. When I change it to globally readable/writeable (0666 root root), `goodhosts` succeeds.

**Comment by cfergeau** [comment]
The `goodhosts` binary should be suid so that it can be used with a `0644 root root` `/etc/hosts` so there's something unexpected going on here.

**Comment by Christopher-Chianelli** [comment]
`goodhosts` is `setuid`, but it still cannot access the `/etc/hosts` file:

```
[cchianel@ada bin]$ ls -l /etc/hosts
-rw-r--r--. 1 root root 321 Jul 29 11:22 /etc/hosts
[cchianel@ada bin]$ ls -l goodhosts 
-r-s--x---. 1 root cchianel 5551059 Jul 28 11:12 goodhosts
[cchianel@ada bin]$ ./goodhosts list
Host file not writable. Try running with elevated privileges.
[cchianel@ada bin]$ ./goodhosts -d list
DEBU[0000] loading default hosts file: /etc/hosts       
Host file not writable. Try running with elevated privileges.
[cchianel@ada bin]$ sudo ./goodhosts list
[sudo] password for cchianel: 
INFO[0000] 127.0.0.1   localhost localhost.localdomain localhost4 localhost4.localdomain4 
INFO[0000] ::1         localhost localhost.localdomain localhost6 localhost6.localdomain6 
INFO[0000] {IP Address} api.crc.testing oauth-openshift.apps-crc.testing console-openshift-console.apps-crc.testing default-route-openshift-image-registry.apps-crc.testing
```

**Comment by cfergeau** [comment]
I was able to reproduce on my setup, however after playing with goodhosts permissions/ownership, the issue was gone, and even after setting the same permissions/owner/group on goodhosts,  I can no longer reproduce, so I don't know what was going on :(

**Comment by praveenkumar** [comment]
closing this issue @Christopher-Chianelli I hope it is fixed for you also?

**Comment by Magiverfox31** [comment]
Hello, guys,

How to solve the issue in windows?
"{{ Error running post start: host file not writable, try running with elevated privileges }}"

**Comment by cfergeau** [comment]
@Magiverfox31 can you open a new issue for this, it's likely unrelated to this issue which was linux specific.

**Comment by cfergeau** [comment]
Any idea what the permissions of /etc/hosts were before you changed them?

**Comment by c7b3rx** [comment]
It's still not fixed :(  I had to apply the workaround to make it work
